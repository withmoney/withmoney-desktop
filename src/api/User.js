import axios from 'axios';

export const login = data => axios.post('http://localhost:3000/api/v1/login', data);

export const signup = data => axios.post('http://localhost:3000/api/v1/signup', data);
