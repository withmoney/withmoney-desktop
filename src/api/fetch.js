import axios from 'axios';
import { catchInvalidToken } from 'app/utils/parse';

const newAxios = axios.create({
  baseURL: `${process.env.MYMONEY_API}`,
  headers: {
    Authorization: {
      toString() {
        return `Bearer ${localStorage.getItem('token')}`;
      },
    },
  },
});

const fetch = method => (...args) => catchInvalidToken(newAxios[method](...args));

export default {
  get: fetch('get'),
  post: fetch('post'),
  put: fetch('put'),
  delete: fetch('delete'),
};
