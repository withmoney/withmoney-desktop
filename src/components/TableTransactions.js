import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import classname from 'classnames';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';
import ButtonRounded from './ButtonRounded';
import * as TransactionsActions from '../store/modules/transactions';
import TransactionsList from './TableTransactionList';

class TableTransactions extends React.Component {
  constructor(props) {
    super(props);

    this.changeTab = this.changeTab.bind(this);
    this.onNextMonth = this.onNextMonth.bind(this);
    this.onPreviousMonth = this.onPreviousMonth.bind(this);

    const currentMoment = moment.tz('UTC');

    this.state = {
      type: 'in',
      currentMonth: currentMoment,
      previousMonth: currentMoment.clone().subtract(1, 'month'),
      nextMonth: currentMoment.clone().add(1, 'month'),
    };
  }

  componentDidMount() {
    this.getTransactions();
  }

  onNextMonth() {
    const { currentMonth } = this.state;

    this.setState({
      currentMonth: moment(currentMonth.add(1, 'month')),
      previousMonth: moment(currentMonth.subtract(1, 'month')),
      nextMonth: moment(currentMonth.add(2, 'month')),
    }, () => {
      this.getTransactions();
    });
  }

  onPreviousMonth() {
    const { currentMonth } = this.state;

    this.setState({
      currentMonth: moment(currentMonth.subtract(1, 'month')),
      previousMonth: moment(currentMonth.subtract(1, 'month')),
      nextMonth: moment(currentMonth.add(2, 'month')),
    }, () => {
      this.getTransactions();
    });
  }

  async getTransactions() {
    const { actions } = this.props;
    const { type, currentMonth } = this.state;

    const start = moment(currentMonth).startOf('month').toISOString();
    const end = moment(currentMonth).endOf('month').toISOString();

    const query = {
      batch: 'Categories',
      order: 'transactionDate.asc',
      type,
      transactionDate: [
        start,
        end,
      ].join(','),
    };

    await actions.transactions.list(query);
  }

  changeTab(typeTab, type) {
    this.setState({
      [typeTab]: type,
    }, () => {
      this.getTransactions();
    });
  }

  render() {
    const {
      type,
      currentMonth,
      nextMonth,
      previousMonth,
    } = this.state;
    const { transactions } = this.props;

    return (
      <div className="table-transactions">
        <div className="table-transactions__tabs">
          <div className="tab-in-out">
            <button
              type="button"
              id="tab-in"
              className={classname(
                'tab-in-out__tab',
                { 'tab-in-out__tab--in-actived': type === 'in' },
              )}
              onClick={() => this.changeTab('type', 'in')}
            >
              In
            </button>
            <button
              type="button"
              id="tab-out"
              className={classname(
                'tab-in-out__tab',
                { 'tab-in-out__tab--out-actived': type === 'out' },
              )}
              onClick={() => this.changeTab('type', 'out')}
            >
              Out
            </button>
          </div>
          <div className="tab-months">
            <button type="button" className="tab-months__navigation" onClick={this.onPreviousMonth}>{previousMonth.format('MMMM YY')}</button>
            <span className="tab-months__current">{currentMonth.format('MMMM YY')}</span>
            <button type="button" className="tab-months__navigation" onClick={this.onNextMonth}>{nextMonth.format('MMMM YY')}</button>
          </div>
        </div>
        <div className="table-transactions__header">
          <div className="table-transactions__header-col">Date</div>
          <div className="table-transactions__header-col">Name</div>
          <div className="table-transactions__header-col">Category</div>
          <div className="table-transactions__header-col">Value</div>
          <div className="table-transactions__header-col">Is Paid?</div>
        </div>
        <div className="table-transactions__body">
          <TransactionsList list={transactions.data} />
          <div className="table-transactions__action">
            <ButtonRounded>Add Transaction</ButtonRounded>
          </div>
        </div>
        <div className="table-transactions__footer">
          <div className="table-transactions__labels">
            <div className="table-transactions__label">All Outs</div>
            <div className="table-transactions__label">All In</div>
            <div className="table-transactions__label">Total</div>
          </div>
          <div className="table-transactions__values">
            <div className="table-transactions__value">$ 5,000.12</div>
            <div className="table-transactions__value">$ 5,000.12</div>
            <div className="table-transactions__value">$ 5,000.12</div>
          </div>
        </div>
      </div>
    );
  }
}

TableTransactions.propTypes = {
  actions: PropTypes.shape({
    transactions: PropTypes.shape({
      list: PropTypes.func.isRequired,
    }).isRequired,
  }).isRequired,
  transactions: PropTypes.shape({
    data: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      transactionDate: PropTypes.string,
      name: PropTypes.string,
      CategoryId: PropTypes.number,
      value: PropTypes.string,
    })),
  }),
};

TableTransactions.defaultProps = {
  transactions: {
    data: [],
  },
};

const mapDispatchToProps = dispatch => ({
  actions: {
    transactions: bindActionCreators(TransactionsActions, dispatch),
  },
});

const mapStateToProps = ({ transactions }) => ({ transactions });

export default connect(mapStateToProps, mapDispatchToProps)(TableTransactions);
