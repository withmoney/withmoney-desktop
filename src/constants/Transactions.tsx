import { TransactionType } from '../models';

export const addOperationText = {
  [TransactionType.Deposit]: 'Add Entrance',
  [TransactionType.FixedExpense]: 'Add Recurrent',
  [TransactionType.CreditCard]: 'Add Credit',
  [TransactionType.VariableExpense]: 'Add Unforeseen',
};

export const operationType = [
  [TransactionType.Deposit],
  [TransactionType.FixedExpense],
  [TransactionType.CreditCard],
  [TransactionType.VariableExpense],
];
