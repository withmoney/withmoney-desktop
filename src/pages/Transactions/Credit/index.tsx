import React from 'react';
import Container from '../Components/Container';
import { Buttons } from '../Components/Buttons';

const Credit = () => {
  return (
    <Container>
      <Buttons activeButton="Credit" />
    </Container>
  );
};

export default Credit;
