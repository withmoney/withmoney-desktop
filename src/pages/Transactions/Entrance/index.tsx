import React from 'react';
import styled from 'styled-components';
import { Buttons } from '../Components/Buttons';
import { Table, TableConfig } from '../Components/Table';
import Container from '../Components/Container';
import EntranceButton from './EntranceButton';
import EntranceTitle from './EntranceTitle';
import EntranceData from './EntranceData';
import EntranceLoad from './EntranceLoad';
import EntranceInfo from './EntranceInfo';

const Entrance = () => {
  return (
    <Container>
      <Buttons activeButton="Entrance" />
      <OperationContainer>
        <Table>
          <TableConfig />
          <EntranceTitle />
          <EntranceData />
          <EntranceLoad />
        </Table>
        <EntranceButton />
      </OperationContainer>
      <EntranceInfo />
    </Container>
  );
};

const OperationContainer = styled.div`
  display: flex;
  margin-bottom: 15px;
  flex-direction: column;
  padding: 30px;
  background-color: #ffff;
`;

export default Entrance;
