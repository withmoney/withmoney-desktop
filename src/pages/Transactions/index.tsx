import React from 'react';
import styled from 'styled-components';
import { Plus } from '@styled-icons/evaicons-solid';
import { Tabs } from './Components/Tabs';
import Table from './Components/Table';
import Button from '../../components/Button';
import Container from './Components/style/Container.style';
import EntranceTitle from './Operation/OperationTitle';
import EntranceData from './Operation/OperationData';
import EntranceLoad from './Operation/TransactionPlaceholder';
import FooterContainer from './Operation/FooterContainer';

const Operation = () => {
  return (
    <Container>
      <Tabs />
      <OperationContainer>
        <Table.Table>
          <Table.Config />
          <EntranceTitle />
          <EntranceData />
          <EntranceLoad />
        </Table.Table>
        <ButtonContent>
          <Button rounded variation="light">
            <Plus />
            <span>Add Entrance</span>
          </Button>
        </ButtonContent>
      </OperationContainer>
      <FooterContainer />
    </Container>
  );
};

const OperationContainer = styled.div`
  display: flex;
  margin-bottom: 15px;
  flex-direction: column;
  padding: 30px;
  background-color: #ffff;
`;

const ButtonContent = styled.div`
  display: flex;
  justify-content: center;
`;

export default Operation;
