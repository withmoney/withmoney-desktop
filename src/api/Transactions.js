import fetch from 'api/fetch';
import { mountQuery } from 'app/utils/parse';

export const list = (query = {}) => fetch.get(mountQuery('transactions', query));

export const get = id => fetch.get(`transactions/${id}`);

export const put = (id, data) => fetch.put(`transactions/${id}`, data);
