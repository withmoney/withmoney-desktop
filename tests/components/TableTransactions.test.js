/* eslint-disable prefer-destructuring */
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import TableTransactions from 'components/TableTransactions';
import * as Transactions from 'api/Transactions';

const mockStore = configureStore([thunk]);

jest.mock('api/Transactions', () => ({
  list: jest.fn().mockResolvedValue({}),
}));
jest.mock('components/TableTransactionList', () => 'table-transaction-list');

describe('TableTransactions', () => {
  let now;
  let Component;

  beforeAll(() => {
    Component = global.withRedux(mockStore({}), TableTransactions);
    now = Date.now;
    Date.now = () => 1548860400000;
  });

  afterAll(() => {
    Date.now = now;
  });

  it('should render with any one transaction entity', () => {
    Transactions.list.mockResolvedValueOnce({
      data: [],
    });

    const wrapper = renderer.create(Component());

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render table with list of transactions', () => {
    Transactions.list.mockResolvedValueOnce({
      data: [],
    });

    const ComponentWithData = global.withRedux(
      mockStore({
        transactions: {
          isLoading: false,
          data: [
            {
              id: 99,
              name: 'Name',
              transactionDate: '2018-08-01 20:80',
              CategoryId: 1,
              type: 'in',
              value: '10.00',
            },
            {
              id: 101,
              name: 'Name',
              transactionDate: '2018-08-01 20:80',
              CategoryId: 1,
              type: 'out',
              value: '20.00',
            },
          ],
        },
      }),
      TableTransactions,
    );

    const wrapper = renderer.create(ComponentWithData());

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should call fetch list on componentDidMount', () => {
    Date.now = () => 1532473578215;
    const wrapper = mount(Component());

    expect(Transactions.list).toBeCalledWith({
      batch: 'Categories',
      order: 'transactionDate.asc',
      transactionDate: '2018-07-01T00:00:00.000Z,2018-07-31T23:59:59.999Z',
    });

    wrapper.find('#tab-out').simulate('click');

    expect(Transactions.list).toBeCalledWith({
      batch: 'Categories',
      order: 'transactionDate.asc',
      transactionDate: '2018-07-01T00:00:00.000Z,2018-07-31T23:59:59.999Z',
    });

    wrapper.find('#tab-in').simulate('click');

    expect(Transactions.list).toBeCalledWith({
      batch: 'Categories',
      order: 'transactionDate.asc',
      transactionDate: '2018-07-01T00:00:00.000Z,2018-07-31T23:59:59.999Z',
    });
  });
});
