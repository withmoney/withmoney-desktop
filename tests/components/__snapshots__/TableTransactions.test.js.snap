// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableTransactions should render table with list of transactions 1`] = `
<div
  className="table-transactions"
>
  <div
    className="table-transactions__tabs"
  >
    <div
      className="tab-in-out"
    >
      <button
        className="tab-in-out__tab tab-in-out__tab--in-actived"
        id="tab-in"
        onClick={[Function]}
        type="button"
      >
        In
      </button>
      <button
        className="tab-in-out__tab"
        id="tab-out"
        onClick={[Function]}
        type="button"
      >
        Out
      </button>
    </div>
    <div
      className="tab-months"
    >
      <button
        className="tab-months__navigation"
        onClick={[Function]}
        type="button"
      >
        December 18
      </button>
      <span
        className="tab-months__current"
      >
        January 19
      </span>
      <button
        className="tab-months__navigation"
        onClick={[Function]}
        type="button"
      >
        February 19
      </button>
    </div>
  </div>
  <div
    className="table-transactions__header"
  >
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Is Paid?
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Date
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "20%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Name
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "20%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Category
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Value
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Action
      </div>
    </div>
  </div>
  <div
    className="table-transactions__body"
  >
    <table-transaction-list
      columns={
        Array [
          Object {
            "label": "Is Paid?",
            "name": "isPaid",
            "style": Object {
              "flexBasis": "15%",
            },
          },
          Object {
            "label": "Date",
            "name": "transactionDate",
            "style": Object {
              "flexBasis": "15%",
            },
          },
          Object {
            "label": "Name",
            "name": "name",
            "style": Object {
              "flexBasis": "20%",
            },
          },
          Object {
            "label": "Category",
            "name": "CategoryId",
            "style": Object {
              "flexBasis": "20%",
            },
          },
          Object {
            "label": "Value",
            "name": "value",
            "style": Object {
              "flexBasis": "15%",
            },
          },
          Object {
            "label": "Action",
            "name": "action",
            "style": Object {
              "flexBasis": "15%",
            },
          },
        ]
      }
      isLoading={false}
      list={
        Array [
          Object {
            "CategoryId": 1,
            "id": 99,
            "name": "Name",
            "transactionDate": "2018-08-01 20:80",
            "type": "in",
            "value": "10.00",
          },
        ]
      }
    />
    <div
      className="table-transactions__action"
    >
      <button
        className="button-rounded"
        disabled={false}
        type="button"
      >
        Add Transaction
      </button>
    </div>
  </div>
  <div
    className="table-transactions__footer"
  >
    <div
      className="table-transactions__labels"
    >
      <div
        className="table-transactions__label"
      >
        All Outs
      </div>
      <div
        className="table-transactions__label"
      >
        All In
      </div>
      <div
        className="table-transactions__label"
      >
        Total
      </div>
    </div>
    <div
      className="table-transactions__values"
    >
      <div
        className="table-transactions__value"
      >
        $10.00
      </div>
      <div
        className="table-transactions__value"
      >
        $20.00
      </div>
      <div
        className="table-transactions__value"
      >
        $30.00
      </div>
    </div>
  </div>
</div>
`;

exports[`TableTransactions should render with any one transaction entity 1`] = `
<div
  className="table-transactions"
>
  <div
    className="table-transactions__tabs"
  >
    <div
      className="tab-in-out"
    >
      <button
        className="tab-in-out__tab tab-in-out__tab--in-actived"
        id="tab-in"
        onClick={[Function]}
        type="button"
      >
        In
      </button>
      <button
        className="tab-in-out__tab"
        id="tab-out"
        onClick={[Function]}
        type="button"
      >
        Out
      </button>
    </div>
    <div
      className="tab-months"
    >
      <button
        className="tab-months__navigation"
        onClick={[Function]}
        type="button"
      >
        December 18
      </button>
      <span
        className="tab-months__current"
      >
        January 19
      </span>
      <button
        className="tab-months__navigation"
        onClick={[Function]}
        type="button"
      >
        February 19
      </button>
    </div>
  </div>
  <div
    className="table-transactions__header"
  >
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Is Paid?
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Date
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "20%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Name
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "20%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Category
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Value
      </div>
    </div>
    <div
      className="table-transactions__header-col"
      style={
        Object {
          "flexBasis": "15%",
        }
      }
    >
      <div
        className="table-transactions__header-col-inner"
      >
        Action
      </div>
    </div>
  </div>
  <div
    className="table-transactions__body"
  >
    <table-transaction-list
      columns={
        Array [
          Object {
            "label": "Is Paid?",
            "name": "isPaid",
            "style": Object {
              "flexBasis": "15%",
            },
          },
          Object {
            "label": "Date",
            "name": "transactionDate",
            "style": Object {
              "flexBasis": "15%",
            },
          },
          Object {
            "label": "Name",
            "name": "name",
            "style": Object {
              "flexBasis": "20%",
            },
          },
          Object {
            "label": "Category",
            "name": "CategoryId",
            "style": Object {
              "flexBasis": "20%",
            },
          },
          Object {
            "label": "Value",
            "name": "value",
            "style": Object {
              "flexBasis": "15%",
            },
          },
          Object {
            "label": "Action",
            "name": "action",
            "style": Object {
              "flexBasis": "15%",
            },
          },
        ]
      }
      isLoading={false}
      list={Array []}
    />
    <div
      className="table-transactions__action"
    >
      <button
        className="button-rounded"
        disabled={false}
        type="button"
      >
        Add Transaction
      </button>
    </div>
  </div>
</div>
`;
